File: src/main/java/com/greenops/workflowtrigger/api/model/GitRepoSchema.java
package com.greenops.workflowtrigger.api.model;

import org.json.JSONException;
import org.json.JSONObject;

public class GitRepoSchema {
    // TODO: The Git repo connection should start with machine accounts (usrname/pass), but then extend to other
    //  methods...this will probably become an interface
    // TODO: Make a test class for GitRepoSchema

    private static String gitRepoName = "Git Repo";
    private static String rootName = "Root";
    private static String gitCredName = "Git Cred";

    private String gitRepo;
    private String pathToRoot;
    private GitCred gitCred;

    public GitRepoSchema(String gitRepo, String pathToRoot, GitCred gitCred) {
        this.gitRepo = gitRepo;
        this.pathToRoot = pathToRoot;
        this.gitCred = gitCred;
    }

    public void setGitRepo(String gitRepo) {
        this.gitRepo = gitRepo;
    }

    public void setPathToRoot(String pathToRoot) {
        this.pathToRoot = pathToRoot;
    }

    public void setGitCred(GitCred gitCred) {
        this.gitCred = gitCred;
    }

    public String getGitRepo() {
        return gitRepo;
    }

    public String getPathToRoot() {
        return pathToRoot;
    }

    public JSONObject convertToJson() {
        try {
            return new JSONObject().put(gitRepoName, gitRepo)
                    .put(rootName, pathToRoot)
                    .put(gitCredName, gitCred.convertToJson());
        } catch (JSONException e) {
            throw new RuntimeException("Git repo information could not be stored as a JSON object");
        }
    }
}

File: workflowtrigger/util/git/git_repo_schema.go
package git

import (
	"encoding/json"
)

const (
	gitRepoName string = "Git Repo"
	rootName    string = "Root"
	gitCredName string = "Git Cred"
)

type GitRepoSchema struct {
	GitRepo    string  `json:"gitRepo"`
	PathToRoot string  `json:"pathToRoot"`
	GitCred    GitCred `json:"gitCred"`
}

func New(gitRepo string, pathToRoot string, gitCred GitCred) GitRepoSchema {
	return GitRepoSchema{
		GitRepo:    gitRepo,
		PathToRoot: pathToRoot,
		GitCred:    gitCred,
	}
}

func (g *GitRepoSchema) SetGitRepo(gitRepo string) {
	g.GitRepo = gitRepo
}

func (g *GitRepoSchema) SetPathToRoot(pathToRoot string) {
	g.PathToRoot = pathToRoot
}

func (g *GitRepoSchema) SetGitCred(gitCred GitCred) {
	g.GitCred = gitCred
}

func (g *GitRepoSchema) GetGitRepo() string {
	return g.GitRepo
}

func (g *GitRepoSchema) GetPathToRoot() string {
	return g.PathToRoot
}

func (g *GitRepoSchema) GetGitCred() GitCred {
	return g.GitCred
}

func UnmarshallGitRepoSchema(m map[string]interface{}) GitRepoSchema {
	gitCred := UnmarshallGitCred(m["gitCred"].(map[string]interface{}))
	return GitRepoSchema{
		GitRepo:    m["gitRepo"].(string),
		PathToRoot: m["pathToRoot"].(string),
		GitCred:    gitCred,
	}
}

func UnmarshallGitRepoSchemaString(str string) GitRepoSchema {
	var m map[string]interface{}
	_ = json.Unmarshal([]byte(str), &m)
	return UnmarshallGitRepoSchema(m)
}

func MarshalGitRepoSchema(schema GitRepoSchema) map[string]interface{} {
	bytes, err := json.Marshal(schema)
	if err != nil {
		panic(err)
	}
	var mapObj map[string]interface{}
	err = json.Unmarshal(bytes, &mapObj)
	if err != nil {
		panic(err)
	}
	mapObj["gitCred"] = MarshalGitCred(schema.GetGitCred())
	return mapObj
}